packages/nfts/contracts/trailblazers-badges/ECDSAWhitelist.sol

```diff
@@ -0,0 +1,150 @@
+/// @dev Signature-driven mint whitelist
+/// @custom:security-contact security@taiko.xyz
+contract ECDSAWhitelist is ContextUpgradeable, UUPSUpgradeable, Ownable2StepUpgradeable {
+    event MintSignerUpdated(address _mintSigner);
+    event MintConsumed(address _minter, uint256 _tokenId);
+    event BlacklistUpdated(address _blacklist);
+
+    error MINTS_EXCEEDED();
+    error ADDRESS_BLACKLISTED();
+    error ONLY_MINT_SIGNER();
+
+    /// @notice Mint signer address
+    address public mintSigner;
+    /// @notice Tracker for minted signatures
+    mapping(bytes signature => bool hasMinted) public minted;
+    /// @notice Blackist address
+    IMinimalBlacklist public blacklist;
+    /// @notice Gap for upgrade safety
+    uint256[47] private __gap;
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    /// @notice Modifier to restrict access to the mint signer
+    modifier onlyMintSigner() {
+        if (msg.sender != mintSigner) revert ONLY_MINT_SIGNER();
+        _;
+    }
+
+    /// @notice Update the blacklist address
+    /// @param _blacklist The new blacklist address
+    function updateBlacklist(IMinimalBlacklist _blacklist) external onlyOwner {
+        blacklist = _blacklist;
+        emit BlacklistUpdated(address(_blacklist));
+    }
+
+    /// @notice Update the mint signer address
+    /// @param _mintSigner The new mint signer address
+    function updateMintSigner(address _mintSigner) public onlyOwner {
+        mintSigner = _mintSigner;
+        emit MintSignerUpdated(_mintSigner);
+    }
+
+    /// @notice Contract initializer
+    /// @param _owner Contract owner
+    /// @param _mintSigner Mint signer address
+    /// @param _blacklist Blacklist address
+    function initialize(
+        address _owner,
+        address _mintSigner,
+        IMinimalBlacklist _blacklist
+    )
+        external
+        initializer
+    {
+        __ECDSAWhitelist_init(_owner, _mintSigner, _blacklist);
+    }
+
+    /// @notice Generate a standardized hash for externally signing
+    /// @param _minter Address of the minter
+    /// @param _tokenId ID for the token to mint
+    function getHash(address _minter, uint256 _tokenId) public pure returns (bytes32) {
+        return keccak256(bytes.concat(keccak256(abi.encode(_minter, _tokenId))));
+    }
+
+    /// @notice Internal method to verify valid signatures
+    /// @param _signature Signature to verify
+    /// @param _minter Address of the minter
+    /// @param _tokenId ID for the token to mint
+    /// @return Whether the signature is valid
+    function _isSignatureValid(
+        bytes memory _signature,
+        address _minter,
+        uint256 _tokenId
+    )
+        internal
+        view
+        returns (bool)
+    {
+        bytes32 _hash = getHash(_minter, _tokenId);
+        (address _recovered,,) = ECDSA.tryRecover(_hash, _signature);
+
+        return _recovered == mintSigner;
```

Comment:
Why do we need the `_minter` input then? Why not simply use `bytes32 _hash = bytes32(_tokenId)`;

Reply:
The `_minter` parameter is the end wallet that wants to mint. `mintSigner` needs to sign the requests specifically for that wallet, otherwise anyone could mint with a valid signature, regardless of their address.